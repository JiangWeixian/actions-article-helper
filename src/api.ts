import { ChatGPTAPI } from 'chatgpt'
import { fetch } from 'cross-fetch'

globalThis.fetch = fetch

export const createChatGPTAPI = (apiKey: string) => {
  return new ChatGPTAPI({
    apiKey,
  })
}
