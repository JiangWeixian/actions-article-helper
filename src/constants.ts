import { template } from 'lodash-es'

export const DEBUG_KEY = 'neo:article-helper'
export const COMMENT_AUTHOR = new Set(['github-actions[bot]'])
export const prefix = '<!--article-helper-->'
export const prompts = {
  // refs: https://github.com/f/awesome-chatgpt-prompts#act-as-an-english-translator-and-improver
  check: template('I want you to act as an English translator, spelling corrector and improver. I will speak to you in any language and you will detect the language, translate it and answer in the corrected and improved version of my text, in English. I want you to replace my simplified A0-level words and sentences with more beautiful and elegant, upper level English words and sentences. Keep the meaning same, but make them more literary. I want you to only reply the correction, the improvements and nothing else, do not write explanations. My markdown format article is "<%= content %>"'),
  // refs: https://wfhbrian.com/the-best-way-to-summarize-a-paragraph-using-gpt-3/
  summarize: template('Write short introduction with high source compression, removes stop words and summarizes the article whilst retaining meaning. The result is the shortest possible summary that retains all of the original meaning and context of the article.'),
}
export const codeBlockRE = /`{3}[\s\S]*?`{3}/g
