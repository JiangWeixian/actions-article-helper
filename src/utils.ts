import { get_encoding } from '@dqbd/tiktoken'
import { title } from 'functional-md'

import { codeBlockRE, prefix } from './constants'

const tokenizer = get_encoding('cl100k_base')

function encode(input: string) {
  return tokenizer.encode(input)
}

// refs: https://github.com/transitive-bullshit/chatgpt-api/blob/0ed1ad003568eba776da2dbd78a4edc73f8acad3/src/chatgpt-api.ts#L410
export const getTokenCount = async (text: string) => {
  text = text.replace(/<\|endoftext\|>/g, '')
  return encode(text).length
}

/**
 * @description Preprocess markdown content, e.g.
 * - remove code blocks
 */
export const preprocess = (content: string, { removeCodeblocks = false }: { removeCodeblocks: boolean } = { removeCodeblocks: false }) => {
  let resolvedContent = content
  if (removeCodeblocks) {
    resolvedContent = resolvedContent.replace(codeBlockRE, '```\n```')
  }
  return resolvedContent
}

const withLeadPrefix = (body: string) => {
  return `${prefix}\n${new Date()}\n${body}`
}

export const formatComment = ({ article, summary }: { article: string; summary: string }) => {
  const comment = `
  ${title({ heading: 1, children: 'Summary' })}
  ${summary}
  ${title({ heading: 1, children: 'Article' })}
  ${article}
  `
  return withLeadPrefix(comment)
}
