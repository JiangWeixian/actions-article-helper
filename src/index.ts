import * as core from '@actions/core'
import * as github from '@actions/github'
import Debug from 'debug'
import matter from 'gray-matter'

import { createChatGPTAPI } from './api'
import { COMMENT_AUTHOR, DEBUG_KEY, prefix, prompts } from './constants'
import { formatComment } from './utils'

const debug = Debug(DEBUG_KEY)

export interface Comment {
  id: number
  body?: string
  user: {
    login: string
  } | null
  created_at: string
}

const findComment = (comments: Comment[]) => {
  return comments.find(comment => comment.user?.login && COMMENT_AUTHOR.has(comment.user?.login) && comment.body?.includes(prefix))
}

const parseArticle = (body: string) => {
  const meta = matter(body, { delimiters: core.getInput('delimiters') ? JSON.parse(core.getInput('delimiters')) : undefined })
  return {
    ...meta,
    data: meta.data,
  }
}

async function main() {
  try {
    debug('start actions')
    const token = process.env.GITHUB_TOKEN
    const apiKey = process.env.OPENAI_API_KEY
    if (!token || !apiKey) {
      throw new Error('Github token and open api key <https://platform.openai.com/account/api-keys> is required.')
    }
    const octokit = github.getOctokit(token)
    const ownerOnly = core.getInput('ownerOnly') ?? true
    debug('Option: ownerOnly %s', ownerOnly)
    const { owner, repo } = github.context.repo
    const { number, owner: issueOwner } = github.context.issue
    if (issueOwner !== owner && ownerOnly) {
      throw new Error('Only owner can trigger this action.')
    }
    debug('issue context %O', github.context.issue)
    const issue = await octokit.rest.issues.get({
      issue_number: number,
      owner,
      repo,
    })
    if (!issue.data.body) {
      core.info('issue body is empty, skip.')
      return
    }
    const issueBody = parseArticle(issue.data.body!)
    const articlePrompt = prompts.check({ content: issueBody.content })
    const chatgptApi = await createChatGPTAPI(apiKey, { article: articlePrompt })
    debug('start check grammar...')
    const articleResult = await chatgptApi.sendMessage(articlePrompt, {
      stream: true,
    })
    debug('chatgpt response for grammar checker %o', articleResult)
    debug('start summarize article...')
    const summaryPrompt = prompts.summarize()
    const summaryResult = await chatgptApi.sendMessage(summaryPrompt, {
      stream: true,
      parentMessageId: articleResult.id,
    })
    debug('chatgpt response for summary %o', summaryResult)
    // list all <number> comments to find specific comments
    const comments = await octokit.rest.issues.listComments({
      owner,
      repo,
      issue_number: number,
    })
    debug('all comments %o from issue %s', comments.data, number)
    // find comments by bot<github_actioins> start with `<!--article-helper-->`
    const comment = findComment(comments.data)
    debug('find comment %o', comment)
    if (comment) {
      await octokit.rest.issues.updateComment({
        owner,
        repo,
        comment_id: comment?.id,
        body: formatComment({ article: articleResult.text, summary: summaryResult.text }),
      })
    } else {
      await octokit.rest.issues.createComment({
        owner,
        repo,
        issue_number: number,
        body: formatComment({ article: articleResult.text, summary: summaryResult.text }),
      })
    }
  } catch (error) {
    core.setFailed((error as any).message)
  }
}

main()
