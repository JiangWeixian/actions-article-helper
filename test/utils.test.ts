import fs from 'node:fs/promises'
import path from 'node:path'

import { describe, expect, it } from 'vitest'

import { formatComment, getTokenCount, preprocess } from '../src/utils'

it('tokens count', async () => {
  const content = (await fs.readFile(path.resolve(__dirname, './prompts.txt'))).toString('utf-8')
  const count = await getTokenCount(content)
  expect(typeof count).toBe('number')
})

describe('preprocess', () => {
  it('remove code blocks', async () => {
    const content = (await fs.readFile(path.resolve(__dirname, './prompts.txt'))).toString('utf-8')
    const count = await getTokenCount(content)
    const contentWithOutCodeblocks = preprocess(content, { removeCodeblocks: true })
    const countWithoutCodeblocks = await getTokenCount(contentWithOutCodeblocks)
    expect(contentWithOutCodeblocks).toMatchSnapshot()
    expect(countWithoutCodeblocks).toBeLessThan(count)
  })
})

it('format', async () => {
  const comment = formatComment({ article: 'content', summary: 'content' })
  expect(comment.split('\n').slice(2).join('\n')).toMatchSnapshot()
})
