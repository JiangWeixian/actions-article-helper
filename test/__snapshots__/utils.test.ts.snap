// Vitest Snapshot v1

exports[`format 1`] = `
"
  # Summary      
  content
  # Article      
  content
  "
`;

exports[`preprocess > remove code blocks 1`] = `
"I want you to act as an English translator, spelling corrector and improver. I will speak to you in any language and you will detect the language, translate it and answer in the corrected and improved version of my text, in English. I want you to replace my simplified A0-level words and sentences with more beautiful and elegant, upper level English words and sentences. Keep the meaning same, but make them more literary. I want you to only reply the correction, the improvements and nothing else, do not write explanations. My markdown format article content is \\"<!--
cover: https://user-images.githubusercontent.com/6839576/226186004-a5fd1446-be38-44ee-82af-44149e3ba9a5.png
description: How to use ChatGPT as a tool for checking grammar and write summary in articles written on GitHub issues.
-->

My blog site is built on top of GitHub issues. When I create or edit my issue on GitHub, \`next.js\` ISR will recreate my blog site in the background. Additionally, ChatGPT is an excellent AI assistant tool that can be used for many purposes, such as a search engine and language translator. As a non-native English speaker, I plan to add ChatGPT to my article writing workflow this weekend to check my articles' grammar like a native English speaker.

## TL;DR

If you write articles on GitHub issues like me, you can use the https://github.com/JiangWeixian/actions-article-helper action in your blog repository.

\`\`\`
\`\`\`

DEMO:

![image](https://user-images.githubusercontent.com/6839576/218319909-bc601e3d-942c-4c05-b49a-4fbab2057b23.png)


## ChatGPT Open API

Lucky, ChatGPT provides an open API. First, you should create an \`API_KEY\` on the OpenAI platform (https://platform.openai.com/account/api-keys).

\`\`\`
\`\`\`

> package \`chatgpt\` is a request client wrapper the basic \`open-ai\` endpoint.

Now, you can send you article content(from issue body) to \`ChatGPT\`, let \`ChatGPT\` to check grammar and rewrite it.

\`\`\`
\`\`\`

By default, you can't send article content like \`Hey, chatgpt help me check {{ content }} grammar\`. Maybe it will not work as you expect.

To make \`ChatGPT\` act as native English speaker, you need **prompts**. **prompts** like language(like \`JavaScript or Rust\`) that \`ChatGPT\` will understand better.

Theres are lots of prompts you can play with from https://github.com/f/awesome-chatgpt-prompts.

In order to make \`ChatGPT\` work as English native. Article content will send with the follow prompts, just replace \`\\"istanbulu cok seviyom burada olmak cok guzel\\"\` with real article content.

> I want you to act as an English translator, spelling corrector and improver. I will speak to you in any language and you will detect the language, translate it and answer in the corrected and improved version of my text, in English. I want you to replace my simplified A0-level words and sentences with more beautiful and elegant, upper level English words and sentences. Keep the meaning same, but make them more literary. I want you to only reply the correction, the improvements and nothing else, do not write explanations. My markdown format article is \\"istanbulu cok seviyom burada olmak cok guzel\\" 

\`\`\`
\`\`\`

You can also use ChatGPT to summary your article after grammar check with follow prompt.

> Write short introduction with high source compression, removes stop words and summarizes the article whilst retaining meaning. The result is the shortest possible summary that retains all of the original meaning and context of the article.

However, if you send this prompt directly. ChatGPT may not know \\"What's your article mentioned inside this prompts\\". Because when you invoke \`sendMessage\`, ChatGPT will create a new convention.

You should ask ChatGPT in previous convention, how to do it? After grammar check, ChatGPT API response contain article results and convention id. Send convention id with summary prompt, now you can get right summary.

![CleanShot 2023-03-20 at 09 59 45@2x](https://user-images.githubusercontent.com/6839576/226229645-793dce1c-1e28-42b9-93b4-5a4280faaebf.png)

The main(hardest) part is complete, now we go further, and make it interaction with github actions.

## Interaction with Github Actions

First, get issue content

\`\`\`
\`\`\`

Then create comment under issue

\`\`\`
\`\`\`

That works great, but not perfect. Before issue publish, issue body will edit multiple times. If you don't wan't create too much comments, I need to find previous comment, and update it.

\`\`\`
\`\`\`\\""
`;
